File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/home/HomeViewModelInteractor.java
package com.dani_chuks.andeladeveloper.starwars.home;


import android.support.annotation.NonNull;

import com.dani_chuks.andeladeveloper.starwars.data.db.repository.FilmRepository;
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.PersonRepository;
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.PlanetRepository;
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.SpecieRepository;
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.StarshipRepository;
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.VehicleRepository;
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Film;
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Person;
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Planet;
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Specie;
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Starship;
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Vehicle;

import java.util.List;

import io.reactivex.Flowable;

public class HomeViewModelInteractor {
    @NonNull private final FilmRepository filmRepository;
    @NonNull private final PlanetRepository planetRepository;
    @NonNull private final SpecieRepository specieRepository;
    @NonNull private final StarshipRepository starshipRepository;
    @NonNull private final VehicleRepository vehicleRepository;
    @NonNull private final PersonRepository personRepository;
    private final static int ITEM_LIMIT = 10;

    public HomeViewModelInteractor(
            @NonNull final FilmRepository filmRepository,
            @NonNull final PlanetRepository planetRepository,
            @NonNull final SpecieRepository specieRepository,
            @NonNull final StarshipRepository starshipRepository,
            @NonNull final VehicleRepository vehicleRepository,
            @NonNull final PersonRepository personRepository) {
        this.filmRepository = filmRepository;
        this.planetRepository = planetRepository;
        this.specieRepository = specieRepository;
        this.starshipRepository = starshipRepository;
        this.vehicleRepository = vehicleRepository;
        this.personRepository = personRepository;
    }

    Flowable<List<Person>> loadPeople(){
       return personRepository.getItemsLimitedToSize(ITEM_LIMIT);
    }

    Flowable<List<Planet>> loadPlanets(){
        return planetRepository.getItemsLimitedToSize(ITEM_LIMIT);
    }

    Flowable<List<Vehicle>> loadVehicles(){
        return vehicleRepository.getItemsLimitedToSize(ITEM_LIMIT);
    }

    Flowable<List<Starship>> loadStarships(){
        return starshipRepository.getItemsLimitedToSize(ITEM_LIMIT);
    }

    Flowable<List<Specie>> loadSpecies(){
        return specieRepository.getItemsLimitedToSize(ITEM_LIMIT);
    }

    Flowable<List<Film>> loadFilms(){
        return filmRepository.getItemsLimitedToSize(ITEM_LIMIT);
    }

}

File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/home/HomeViewModelInteractor.kt
package com.dani_chuks.andeladeveloper.starwars.home


import com.dani_chuks.andeladeveloper.presentation_models.MainModels
import com.dani_chuks.andeladeveloper.presentation_models.mappers.Mapper
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.GetAllBySize
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.film.FilmRepository
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.people.PeopleRepository
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.planet.PlanetRepository
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.specie.SpecieRepository
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.starship.StarshipRepository
import com.dani_chuks.andeladeveloper.starwars.data.db.repository.vehicle.VehicleRepository
import com.dani_chuks.andeladeveloper.starwars.data.models.EntityList
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.*
import com.dani_chuks.andeladeveloper.starwars.di.Result
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject

class HomeViewModelInteractor @Inject constructor(
        val filmRepository: FilmRepository,
        val planetRepository: PlanetRepository,
        val specieRepository: SpecieRepository,
        val starshipRepository: StarshipRepository,
        val vehicleRepository: VehicleRepository,
        val personRepository: PeopleRepository) {

    fun loadPeople(): Flow<List<Person>?> = personRepository.getPeopleByPredicateAsFlow(GetAllBySize(ITEM_LIMIT))

    fun loadPlanets(): Flow<List<Planet>?> = planetRepository.allAsFlow()

    fun loadVehicles(): Flow<List<Vehicle>?> = vehicleRepository.getVehiclesBySizeAsFlow(ITEM_LIMIT)

    fun loadStarships(): Flow<List<MainModels.StarshipModel>?> =
            starshipRepository.getStarShipsLimitedToSizeAsFlow(ITEM_LIMIT)
                    .map { it?.let{ Mapper.mapStarships(it) } }

    fun loadSpecies(): Flow<List<Specie>?> = specieRepository.getItemBySizeAsFlow(ITEM_LIMIT)

    fun loadFilms(): Flow<List<Film>?> = filmRepository.getFilmsByPredicateAsFlow(GetAllBySize(ITEM_LIMIT))

    suspend fun loadFilmsRemote() = filmRepository.fetchAndSync()

    suspend fun loadPeopleRemote(page: Int) = personRepository.fetchAndSync(page)

    suspend fun loadPlanetRemote(page: Int): Result<Boolean> = planetRepository.fetchAndSync(page)

    suspend fun loadVehicleRemote(page: Int) = vehicleRepository.fetchAndSync(page)

    suspend fun loadSpeciesRemote(page: Int) = specieRepository.fetchAndSync(page)

    suspend fun loadStarshipsRemote(page: Int) = starshipRepository.fetchAndSync(page)

    companion object {
        val ITEM_LIMIT = 10
    }

}
